File: client/src/main/java/app/ClientMain.java
package app;

import client.UserConsole;
import collection.CollectionFilter;
import commands.CommandManager;
import commands.CommandNameToInfo;
import commands.pcommands.*;
import communication.*;
import exceptions.ReconnectionTimoutException;
import exceptions.ResponseCodeException;
import model.data.Flat;
import utils.Console;

import java.io.IOException;
import java.io.PrintStream;

public class ClientMain {
  public static void main(String[] args) {
    Session clientSession = new ClientSession("localhost", 2222);
    if (!waitForConnection(clientSession)) {
      System.out.println("Connection failed. Timeout reached.");
      return;
    }

    Transceiver clientTransceiver = new ClientTransceiver(clientSession.getSocketChannel());
    Exchanger<Flat> exchanger = new ClientExchanger<>(clientTransceiver, clientSession);

    CollectionFilter collectionFilter;
    try {
      exchanger.requestFullCollection();
      collectionFilter = new CollectionFilter(exchanger.receiveFullCollection());
    } catch (ReconnectionTimoutException | ResponseCodeException | IOException e) {
      System.err.println("Failed to load collection. Exiting with error: " + e.getMessage());
      return;
    }

    PrintStream writer = new PrintStream(System.out);
    CommandManager clientCommandManager = new CommandManager();
    Console userConsole = new UserConsole(writer, clientCommandManager, exchanger, collectionFilter);

    clientCommandManager.registerCommands(
        new Help(clientCommandManager),
        new Exit(userConsole),
        new ExecuteScript(userConsole),
        new Info(collectionFilter),
        new FilterContainsName(collectionFilter),
        new PrintUniqueNumberOfRooms(collectionFilter),
        new PrintFieldDescendingNew(collectionFilter)
    );

    CommandNameToInfo commandNameToInfo;
    try {
      commandNameToInfo = getAccessibleCommandsInfo(exchanger);
    } catch (ReconnectionTimoutException | ResponseCodeException | IOException e) {
      System.out.println("Can't get accessible commands: " + e.getMessage());
      return;
    }
    clientCommandManager.addCommandInfos(commandNameToInfo);

    userConsole.run();
  }

  private static boolean waitForConnection(Session clientSession) {
    return clientSession.reconnect(15);
  }

  private static CommandNameToInfo getAccessibleCommandsInfo(Exchanger<Flat> exchanger) throws ReconnectionTimoutException, ResponseCodeException, IOException {
    exchanger.requestAccessibleCommandsInfo();
    return exchanger.receiveAccessibleCommandsInfo();
  }
}

File: client-ui/domain/src/commonMain/kotlin/ru.moleus.kollector/domain/app/ClientMain.kt
package ru.moleus.kollector.domain.app

import commands.CommandManagerImpl
import commands.CommandNameToInfo
import common.context.Exchanger
import communication.Transceiver
import ru.moleus.kollector.domain.exceptions.ResponseCodeException
import ru.moleus.kollector.domain.commands.pcommands.*
import ru.moleus.kollector.domain.communication.*
import common.exceptions.InvalidCredentialsException
import ru.moleus.kollector.domain.collection.CollectionFilter
import ru.moleus.kollector.domain.communication.ClientTransceiver
import kotlin.jvm.JvmStatic
import java.io.*

object ClientMain {
    @JvmStatic
    fun main(args: Array<String>) {
        val clientSession: Session = ConnectionSession("localhost", 2222)
        if (!waitForConnection(clientSession)) {
            println("Connection failed. Timeout reached.")
            return
        }
        val clientTransceiver: Transceiver = ClientTransceiver(clientSession.getSocketChannel())
        val exchanger: Exchanger = ClientExchanger(clientTransceiver, clientSession)
        val collectionFilter: CollectionFilter
        try {
            exchanger.requestFullCollection()
            collectionFilter = CollectionFilter(exchanger.receiveFullCollection())
        } catch (e: IOException) {
            println("Failed to load collection. Exiting with error: " + e.message)
            return
        } catch (e: InvalidCredentialsException) {
            println("Failed to load collection. Exiting with error: " + e.message)
            return
        }
        val clientCommandManager = CommandManagerImpl()
//        val userConsole: client.Console = UserConsole(writer, clientCommandManager, exchanger, collectionFilter)
        val authenticator: Authenticator = ClientAuthenticator(exchanger)
        clientCommandManager.registerCommands(
            Help(clientCommandManager),
//            Exit(userConsole),
//            ExecuteScript(userConsole),
            Info(collectionFilter),
            FilterContainsName(collectionFilter),
            PrintUniqueNumberOfRooms(collectionFilter),
            PrintFieldDescendingNew(collectionFilter),
            Login(authenticator),
            Register(authenticator)
        )
        val commandNameToInfo: CommandNameToInfo = try {
            getAccessibleCommandsInfo(exchanger)
        } catch (e: Exception) {
            when(e) {
                is ResponseCodeException, is IOException, is InvalidCredentialsException -> {
                    println("Can't get accessible commands: " + e.message)
                    return
                }
                else -> throw e
            }
        }
        clientCommandManager.addCommandsInfo(commandNameToInfo)
//        userConsole.run()
    }

    private fun waitForConnection(clientSession: Session): Boolean {
        return clientSession.reconnect(15)
    }

    @Throws(ResponseCodeException::class, IOException::class, InvalidCredentialsException::class)
    private fun getAccessibleCommandsInfo(exchanger: Exchanger): CommandNameToInfo {
        exchanger.requestAccessibleCommandsInfo()
        return exchanger.receiveAccessibleCommandsInfo()
    }
}